gnupg-pkcs11-scd/keyutil.c

```diff
@@ -313,3 +435,118 @@ char *keyutil_get_cert_hexgrip (gcry_sexp_t sexp)
 
 	return ret;
 }
+
+void keyinfo_data_free(keyinfo_data_list list) {
+	keyinfo_data_list next, curr;
+
+	if (list == NULL) {
+		return;
+	}
+
+	for (curr = list; curr != NULL; curr = next) {
+		next = curr->next;
+
+		if (curr->value != NULL) {
+			if (curr->value_free != NULL) {
+				curr->value_free(curr->value);
+			}
+		}
+
+		if (curr->tag != NULL) {
+			if (curr->tag_free != NULL) {
+				curr->tag_free(curr->tag);
+			}
+		}
+
+		free(curr);
+	}
+}
+
+keyinfo_data_list keyinfo_get_key_data(keyinfo keyinfo) {
+	keyinfo_data_list first = NULL, n_item = NULL, e_item = NULL;
+	unsigned char *n_hex = NULL;
+	unsigned char *e_hex = NULL;
+
+	if (keyinfo->type == KEYINFO_KEY_TYPE_INVALID) {
+		return NULL;
+	}
+
+	if (keyinfo->type != KEYINFO_KEY_TYPE_UNKNOWN) {
+		return NULL;
+	}
+
+	switch (keyinfo->type) {
+		case KEYINFO_KEY_TYPE_RSA:
+			if (
+				gcry_mpi_aprint (
+					GCRYMPI_FMT_HEX,
+					&n_hex,
+					NULL,
+					keyinfo->data.rsa.n
+				) ||
+				gcry_mpi_aprint (
+					GCRYMPI_FMT_HEX,
+					&e_hex,
+					NULL,
+					keyinfo->data.rsa.e
+				)
+			) {
+				break;
+			}
+
+			e_item = malloc(sizeof(*e_item));
+			if (e_item == NULL) {
+				break;
+			}
+			e_item->next = NULL;
+			e_item->type = (unsigned char *) "KEY-DATA";
+			e_item->tag = (unsigned char *) "e";
+			e_item->value = e_hex;
+			e_item->value_free = gcry_free;
+			e_item->tag_free = NULL;
```

Comment:
won't it be much simpler to just write free explicit logic per key type?

Reply:
In the future, the value may need to be free'd using a different free method.

Reply:
For the EcDSA support the only two values we ever supply to `value_free` are currently `gcry_free` or `NULL` (if the value does not need to be freed).

Reply:
for RSA with have one method of free and ecdsa we have another, why make it that generic?

Reply:
Keeping release near allocation increases the chance that it will get done by people copy/pasting code.

Reply:
In the case where the decision regarding which freeing mechanism (if any) to use is in a different function far from the allocation function, it may get done incorrectly.

Currently the freeing function (`keyinfo_data_free()`) won't need to change as any new key types are added.  In the proposed change, both the freeing function (`keyinfo_data_free()`) and the allocating function (`keyinfo_get_key_data()`) will both need to change in sync.  The default free action must be not to free

Reply:
I understand if we had a very complex application, and in this case we would have a private member of the structure to avoid exposing the internal information, more probably we would not expose a struct as we want to be backward compatible when/if format change and provide accessor methods.

This application is not complex enough or generic infrastructure to require this kind of flexibility.